### ISSUE TITLE
Pipewire backend: Don't set target.object if openal wants to open the default device.

### ISSUE THREAD

**Issue by amano-kenji** [kcat/openal-soft#892]
I just read openal codebase and discovered that it sets target.object even when it wants to open the default device.

On my system, default sink is null-sink which is only used to catch loopback playback streams that miss their corresponding ALSA device nodes.

My pipewire stream.rules have a rule that assigns `default-sink` to `target.object` if `target.object` is missing in a stream.

When telegram desktop uses openal pipewire backend, I cannot hear any notification sound because every sound is redirected to null-sink instead of default-sink.

If there is no target, then it should not set `target.object` to `object.serial` of pipewire's `default.audio.sink` which is null-sink on my system.

**Comment by kcat** [comment]
The problem is, OpenAL Soft needs `default.audio.sink` to know what the default device is, so it will know what to report to the app if/when it queries the default device name. PipeWire also doesn't seem to have a method to query which node a stream is connected to, so if it doesn't already know what to connect to, it has no way to get its name to tell the app what it opened.

It's possible I'm missing something and maybe you can get the default device and a stream's target node some other way, but as it is, this is the only way I know of to make device names work as expected from OpenAL's side.

**Comment by amano-kenji** [comment]
A playback or capture stream's target.object can be queried as below.

* Query the stream's `target.object` from node metadata. On command line, `pw-metadata -n default node.name target.object` reveals `target.object`.
* If `target.object` is not specified in node metadata, query `target.object` in node properties.
* If `target.object` is absent in node properties as well, then there is no `target.object`, and the stream is automatically linked to `configured.default.audio.sink`, `default.audio.sink`, or any next best node.
* If `target.object` is absent, query `node.name` and `node.description` of the node the stream is connected to. `pw-link` can reveal the link.

**Comment by amano-kenji** [comment]
Target can be absent or specified. Even if target is specified, after the target disappears, the stream can be connected to a different node.

Thus, I suggest tracking just target instead of which node a stream is actually connected to. If you want to know the actual node a stream is connected to, you can use `pw-link` or any C API equivalent for `pw-link`.

**Comment by amano-kenji** [comment]
I think pulseaudio backend knows that Default means `target.object` is absent.

**Comment by amano-kenji** [comment]
* Replace `mTargetId`  with `mTargetObject`.

https://docs.pipewire.org/group__pw__stream.html#ga0e12176f13e654e64e7f5689e7a6dd54 says

> target_id should  have the value PW_ID_ANY. To select a specific target node, specify the  PW_KEY_OBJECT_SERIAL or the PW_KEY_NODE_NAME value of the target node  in the PW_KEY_TARGET_OBJECT property of the stream. Specifying target  nodes by their id is deprecated.

* Make sure `Default` in UI means `mTargetObject` is empty.

**Comment by kcat** [comment]
> Don't report the default device name. Just say that `Default` device is any pipewire node that a stream connects to when `target.object` is absent. OpenAL should have a notation for absent `target.object`.

Reporting just "Default" when it's using the default device isn't a good idea, since the purpose of the device name is to say what device is actually being used (if you're trying to troubleshoot what device is being used, or do something different for different outputs, just getting the name "Default" won't be helpful). Knowing what device will be used is also important so it can autoconfigure the output to match, and properly handle device-specific config options.

There's also the issue of the in-progress `ALC_SOFT_system_events` extension that will allow apps to detect when the default device changes, which with PipeWire is done by watching for changes to `default.audio.sink`. If that doesn't actually represent the default device, the app won't be able to respond to the default changing as intended.

> Even if `target.object` is a specific node, after the specific node disappears, the stream is going to be connected to another node automatically by wireplumber.

As long as it's initially accurate, it's not a severe problem if it's changed by external means later. As long as PipeWire sends a notification that it changed, a method can be made to notify the app so it can respond appropriately.

**Comment by amano-kenji** [comment]
`default.audio.*` can often be wrong. Pipewire has a complex logic for selecting nodes.

For example, `default.configured.audio.*` can override `default.audio.*`. Those default nodes may be ignored if the nodes disappear later.

I also have my own custom pipewire/wireplumber rules that ignore `default.audio.*` and `default.configured.audio.*`.

I think openal clients should not care which pipewire node they are connected to if the playback or sink device is `Default` in user interface. `Default` actually means I don't care what the target node is.

OpenAL can choose 2 channel stereo for sink and mono for source if it doesn't know anything about the target. Or, users can configure the default in configuration file?

**Comment by amano-kenji** [comment]
* If `default.configured.audio.*` is set but is pointing to a missing node, refer to `default.audio.*`.
* If `default.audio.*` is set but is pointing to a missing node, pick the node that highest `priority.session`.

This is how default is chosen for nodes that don't have `target.object`.

For streams whose `target.object`s are missing in the pipewire graph, the above procedure is applied to find the default, and wireplumber automatically connects the streams to the dynamically chosen default node.

Thus, if `Default` is chosen, openal should not set `target.object`, and just examine targets of new links that open on the fly. Even if `target.object` is chosen by a user, after `target.object` goes away, openal has to adapt to the new nodes that openal streams are automatically connected to by wireplumber.

**Comment by amano-kenji** [comment]
In my humble opinion, `ALC_SOFT_system_events` should not report when the default node changes. The default can be null-sink or null-source on some systems.

It should report when an OpenAL stream is connected to a new sink/source, and it should automatically adapt channels and sample rate to the new sink/source. `If Default` means `target.object` is not set by OpenAL, pipewire and wireplumber will take over. I want to control things in pipewire and wireplumber.

**Comment by amano-kenji** [comment]
Perhaps, OpenAL should differentiate `target.object` from the actual node a stream is actually connected to.

Then, it's okay to not set `target.object` when target is `Default` in user interface.

**Comment by kcat** [comment]
> In my humble opinion, `ALC_SOFT_system_events` should not report when the default node changes. The default can be null-sink or null-source on some systems.

The app may need to know when the default changes so it can move/reopen the device on the new default, if it wants output to stay with the default as it changes (e.g. plugging in headphones or an HDMI/DP device, and having output move there from wherever it was). Even if the system forcibly moves the stream without the app's input, the app may still need to reopen or reset the device so the configuration can be properly updated. Or it may have its own list and display for available devices, and uses default node change notifications to update itself and show what the default is.

> It should report when an OpenAL stream is connected to a new sink/source, and it should automatically adapt channels and sample rate to the new sink/source regardless of whether OpenAL sets target.object on streams.

It can't adapt during playback since there's no way to safely lock the ALCdevice state to update it, alongside updating the playback stream, from the notification callback. And even if there was a way to, having certain properties randomly change during playback can cause problems for apps that rely on them staying constant, so it's best to only have the configuration change on specific API calls that the app is aware can cause a change.

Adding an event for a device moving to a different node is something that will probably be added, so the app can know the device may need to be reconfigured.

**Comment by amano-kenji** [comment]
If target is `Default`, then it can open a stereo playback stream or a mono capture stream without target, see what is actually connected, close the stream if the connected node is not compatible, create a new compatible stream, and then return the stream to the application. Applications can be given a stream only after this initial negotiation process ends. OpenAL doesn't need to show this negotiation process to the applications. The negotiation can happen transparently.

If users choose a specific target node and the target disappears, wireplumber will connect the stream to another node without changing `target.object`. Perhaps, nothing needs to be done automatically by OpenAL in this case?

**Comment by amano-kenji** [comment]
> Adding an event for a device moving to a different node is something that will probably be added, so the app can know the device may need to be reconfigured.

This event can solve the issue. Whether or not a stream has a user-specified target, a stream can be automatically moved to a new node by wireplumber.

In this event, the application may choose to close the current stream and open a new one. The application may choose automatic negotiation in OpenAL or manually create a new stream compatible with the newly connected node.

**Comment by kcat** [comment]
> see what is actually connected

This is the crux of the issue, finding out what node a stream connects to so OpenAL can know the name and default configuration it should use for the ALCdevice.

**Comment by amano-kenji** [comment]
OpenAL can subscribe to a pipewire event that occurs when a stream is connected to a node. This event will need to be processed asynchronously. Asynchronous programming is a bit tricky, but it can be done with proper design.

**Comment by amano-kenji** [comment]
I started learning pipewire API.

I found state_changed on https://docs.pipewire.org/structpw__stream__events.html.

When the state of a stream transitions to `PW_STREAM_STATE_STREAMING`, you can examine the connected node.

I think `pw-link.c` in pipewire source code can teach someone how to find out which node a stream is connected to.

**Comment by amano-kenji** [comment]
https://docs.pipewire.org/page_streams.html says

> After connecting the stream, the server will want to configure some parameters on the stream. You will be notified of these changes with the param_changed event.

`pw-link.c` provides an example of obtaining the connected node.

**Comment by amano-kenji** [comment]
I tested `mpv`. When `mpv` creates a pipewire stream without target, it automatically detects the node that the stream is connected to. If the connected node is a 5.1 channel node, mpv becomes a 5.1 channel stream. If I move mpv stream to a stereo node by executing `pw-metadata id-of-mpv target.object node.name`, mpv becomes a stereo stream.

https://github.com/mpv-player/mpv/blob/master/audio/out/ao_pipewire.c is how mpv dynamically adapts to the connected nodes on the fly.

**Comment by kcat** [comment]
Adapting to a new sink isn't a problem. The issue is knowing what node a stream is connected to so OpenAL knows what the name of the output/input is, what configuration options are applicable, and other relevant info about the node. Once I can get that, everything will fall into place.

**Comment by amano-kenji** [comment]
You can detect a new connected node by detecting changes to `Latency` parameters and examine the link to find out the new node.

That's one way. I will try to find another way if you are not satisfied.

**Comment by amano-kenji** [comment]
Even though you don't know the name of a new node, you can still adapt the internal audio channels to the new node by examining changes to `PortConfig`.

If you just want to adapt channel configuration, you only need to watch changes to `PortConfig`. What would you need node name and other things for?

**Comment by kcat** [comment]
> Detect changes to `Latency` parameters and examine the link to find out the new node.

I'm not seeing anything in the latency params to get the node being linked to.

> Detect changes to `PortConfig` parameters.

I'm not seeing `SPA_PARAM_PortConfig` come through in the `param_changed` callback. I might be checking wrong since I have no idea how this is supposed to be read, but it doesn't seem to be there. 

> `pw-link.c` in pipewire codebase shows an example of finding out the connected node through the port links.

That seems to do a general probe of all nodes and connections, rather than checking a particular `pw_stream` for the node it connects to. Having to find some way to get the stream's node, then set up a separate registry callback to watch for things happening to that node (and hope I can find the info about what node it connects to), would be annoyingly complicated.

> What would you need node name and other things for?

So I know what the device name is, can get device-specific options, and have access to other info stored about the node.

**Comment by amano-kenji** [comment]
> I'm not seeing SPA_PARAM_PortConfig come through in the param_changed callback. I might be checking wrong since I have no idea how this is supposed to be read, but it doesn't seem to be there.

Have you actually debugged `param_changed` callback on a stream?

I created a new issue on https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/3409 because I am just a newbie who started using pipewire.

**Comment by amano-kenji** [comment]
This is how I got around the issue. Coding the workaround took some learning.

* ~/.config/pipewire/pipewire-pulse.conf.d/10-target.conf
* ~/.config/pipewire/client.conf.d/10-target.conf
* ~/.config/pipewire/client-rt.conf.d/10-target.conf
```
stream.rules = [
  {
    matches = [
      {
        media.class = Stream/Output/Audio
        target.object = null
      }
      {
        media.class = Stream/Output/Audio
        target.object = "null-sink"
      }
    ]
    actions = {
      update-props = {
        target.object = default-sink
      }
    }
  }
  {
    matches = [
      {
        media.class = Stream/Input/Audio
        target.object = null
      }
      {
        media.class = Stream/Input/Audio
        target.object = "null-source"
      }
    ]
    actions = {
      update-props = {
        target.object = default-source
      }
    }
  }
]
```
Since telegram-desktop picks "null-source" with OpenAL pulseaudio backend, the above configuration works for pulseaudio backend.

But, OpenAL pipewire backend picks the `object.serial` of `null-source` or `null-sink`. That's why I came up with this weird wireplumber script. I had to learn some wireplumber programming.

~/.config/wireplumber/main.lua.d/41-stream-defaults.lua
```
stream_defaults.properties["restore-props"]=false
stream_defaults.properties["restore-target"]=false

load_script("null.lua")
```
~/.config/wireplumber/scripts/null.lua
```
allnodes_om = ObjectManager { Interest { type = "node" } }
allnodes_om:activate()

metadata_om = ObjectManager {
  Interest {
    type = "metadata",
    Constraint { "metadata.name", "=", "default" },
  }
}
metadata_om:activate()

streams_om = ObjectManager {
  Interest {
    type = "node",
    Constraint { "media.class", "matches", "Stream/*", type = "pw-global" },
  }
}
streams_om:connect("object-added", function (streams_om, node)
  local null_sink = allnodes_om:lookup{
    Constraint { "node.name", "=", "null-sink", type = "pw-global" }
  }
  local null_source = allnodes_om:lookup{
    Constraint { "node.name", "=", "null-source", type = "pw-global" }
  }

  if null_sink ~= nil and null_source ~= nil
  then
    local null_sink_serial = null_sink.properties["object.serial"]
    local null_source_serial = null_source.properties["object.serial"]

    if node.properties["target.object"] == null_sink_serial
      and node.properties["media.class"] == "Stream/Output/Audio"
    then
      Log.message("A playback stream tried to use serial.object of null-sink as target.object")
      local metadata = metadata_om:lookup()
      metadata:set(node["bound-id"], "target.node", "Spa:String", "default-sink")
    elseif node.properties["target.object"] == null_source_serial
      and node.properties["media.class"] == "Stream/Input/Audio"
    then
      Log.message("A capture stream tried to use serial.object of null-source as target.object")
      local metadata = metadata_om:lookup()
      metadata:set(node["bound-id"], "target.node", "Spa:String", "default-source")
    end
  end
end)
streams_om:activate()
```
`null.lua` replaces object.serial of null-sink with `default-sink` and object.serial of null-source with `default-source`.

**Comment by amano-kenji** [comment]
You may also want to look into https://docs.pipewire.org/group__pw__impl__node.html which has `[pw_impl_node_events](https://docs.pipewire.org/structpw__impl__node__events.html)`.

**Comment by amano-kenji** [comment]
I think [pw_impl_node_events](https://docs.pipewire.org/structpw__impl__node__events.html) may be what you want for this issue.

**Comment by kcat** [comment]
`pw_impl_` doesn't sound like something apps should use, given the `impl`ementation part which usually signifies an internal thing that can change in incompatible ways. Though even if it's safe to use, I'm not sure how to access it from the stream.

**Comment by amano-kenji** [comment]
https://docs.pipewire.org/group__api__pw__impl.html says

> the Implementation API is roughly equivalent to libwayland-server.

According to debian package description, libwayland-server is

> the library that implements the server side of the Wayland protocol.

https://github.com/drewt/chicken-wayland-server is a chicken scheme binding to libwayland-server.

I think it is meant to be used by pipewire clients. Developers would not document internal details that are going to change often.

**Comment by amano-kenji** [comment]
> Though even if it's safe to use, I'm not sure how to access it from the stream, or what kind of properties to look for.

I don't know, either. But, you can learn if you want it enough.

**Comment by amano-kenji** [comment]
With a quick scroll, I found `pw_stream_get_node_id` for getting `node.id` from a stream.

There should be a way to get the node from node id, but I don't know where to get the node, yet.

**Comment by amano-kenji** [comment]
Since a system can change `target.object` by overriding node properties through `stream.rules` or setting `target.object` metadata, I guess an easier way to adapt to an actual target would be

* If a user picks `Default` as the target, leave `target.object` empty. Assume `configured.default.audio.*` is the target. If `configured.default.*` doesn't exist, assume `default.audio.*` is the target. If `default.audio.*` is set, but the corresponding node is missing, then assume the node with the highest `priority.session` is the target.
* Listen to changes to `target.object` metadata. If it is set or changed, adapt.
* If `target.object` meatadata doesn't exist, examine `target.object` stream property because `stream.rules` might have changed it. Adapt to the new `target.object` stream property set by `stream.rules`.

**Comment by ilya-fedin** [comment]
@kcat could pulse backend be prioritized over the pipewire one until this is resolved? This is a major UX issue, users can't use external audio devices connected in runtime. I'd even say it's way more important than being unable to get device name from openal.

**Comment by amano-kenji** [comment]
Someone said most open-source developers are jack-asses who don't care about UI and UX. That's why he says he uses proprietary softwares most of the time.

**Comment by kcat** [comment]
> This is a major UX issue, users can't use external audio devices connected in runtime.

External devices aren't detected/enumerated at all with the PipeWire backend? That sounds like a different and much more serious problem (all devices should be seen, and PipeWire's normal default selection should be recognized). The problem here isn't that devices aren't seen, it's that a user-made script overrides PipeWire device selection behavior so that PipeWire's reported default device doesn't match what a stream auto-selects when set to use "any" target.

Can you get a trace log of when the earbuds are connected? Perhaps they're being reported under an unexpected media class that OpenAL Soft doesn't recognize for playback. Also FWIW, the latest Git versions of OpenAL Soft allow you to get the log yourself, by calling `alcGetProcAddress(NULL, "alsoft_set_log_callback")`, and then calling the retrieved function before any other OpenAL function.

```c++
typedef void (ALC_APIENTRY*LPALSOFTLOGCALLBACK)(void *userptr, char level, const char *message, int length) noexcept;
void ALC_APIENTRY alsoft_set_log_callback(LPALSOFTLOGCALLBACK callback, void *userptr) noexcept;
```

Note however this doesn't work if using the router, since it has to come directly from OpenAL Soft's lib before any OpenAL calls are made on it. And be aware the callback can be called on different threads. But the callback takes a `char level` parameter, which is a character code indicating the severity of the message (currently can be `'E'` for error, `'W'` for warning, and `'I'` for informational; others may show up in the future meaning other things, potentially including critical failures or excessively spammy messages). The message itself is UTF-8, but you shouldn't try to parse it for anything other than adding to your own log, as the text is subject to change (messages may be added, removed, or altered). `const char *message` is null terminated, and `int length` is the length in bytes excluding the null terminator.

**Comment by ilya-fedin** [comment]
Hmm, yeah, this class-based detection seemed weird to me... Could this be done in another way? E.g. look what gstreamer integration is doing to differ between sink and source: https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/gst/gstpipewiredeviceprovider.c#L215-223
And what is doing to differ between video and audio: https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/gst/gstpipewireformat.c#L906-961

**Comment by ilya-fedin** [comment]
Nevermind, the person just seem to have pipewire and pulseaudio running simultaneously.

**Comment by amano-kenji** [comment]
That's right. pipewire can pretend to be a pulseaudio server.

**Comment by amano-kenji** [comment]
And, class-based detection is the way intended by pipewire developers.
